### PR TITLE
Fix internal bridges over Hilbert Curve/Octagram Spiral sparse infill

### PR DESCRIPTION
**Pull request by valerii-bokhan**
# Description

Problem: Some internal bridges can be generated in mid-air. Especially for Hilbert Curve/Octagram Spiral sparse infill.
Fix: Internal bridges are extended until the next available sparse infill wall.

With the help of AI.

# Screenshots

### Before

image

image

image

image

### After

image

image

image

image

Still has rare issues:
image

image

# Test Project

[test_internal_bridges_project.zip](https://github.com/user-attachments/files/32026982/test_internal_bridges_project.zip)

[How to Download Pull Requests Artifacts for Testing](https://www.orcaslicer.com/wiki/how_to_download_pr_artifacts)

### CODE REVIEW

**Comment by RF47** [comment]
I hadn't seen this bug

**Comment by valerii-bokhan** [comment]
> I hadn't seen this bug

I have attached the test project to the main post.

It works better now, but still buggy:

image

image

The fix would be to allow bridges to connect to the next available infill wall, because bridges are currently limited by the area above them.

Something like that:

image

**Comment by valerii-bokhan** [comment]
I have added a fix and updated screenshots in the main post.

File: tests/fff_print/test_printobject.cpp [hunk 1]
```diff
@@ -130,3 +131,102 @@ TEST_CASE("Initial layer height is honored", "[PrintObject]")
...
+    TriangleMesh tower = make_cube(14, 10, 1);
+    tower.translate(8, 7, 3);
+    mesh.merge(tower);
+    return mesh;
+}
+
+static DynamicPrintConfig internal_bridge_config(const std::string &pattern)
+{
+    auto config = DynamicPrintConfig::full_print_config();
+    config.set_deserialize_strict({{"sparse_infill_pattern", pattern},
```
Comment (RF47) [inline tests/fff_print/test_printobject.cpp:153]: it would be good include a multiline count

File: tests/fff_print/test_printobject.cpp [hunk 2]
```diff
@@ -4,6 +4,7 @@
 #include "libslic3r/Print.hpp"
 #include "libslic3r/Layer.hpp"
 #include "libslic3r/GCodeReader.hpp"
+#include "libslic3r/ClipperUtils.hpp"
```
[the remark this answers is not in the document]
Comment (valerii-bokhan) [reply tests/fff_print/test_printobject.cpp:7]: Fixed.

File: src/libslic3r/PrintObject.cpp [hunk 3]
```diff
@@ -1401,15 +1402,16 @@ bool PrintObject::invalidate_state_by_config_options(
             || opt_key == "infill_anchor_max"
             || opt_key == "top_surface_line_width"
             || opt_key == "bottom_surface_density"
-            || opt_key == "center_of_surface_pattern"
```
Comment (RF47) [inline src/libslic3r/PrintObject.cpp:1404]: just in case, Did you check to see if this breaks the “separated infills” for the top and bottom infills?
this was added in #14674 and #11682

Comment (valerii-bokhan) [reply src/libslic3r/PrintObject.cpp:1404]: Let me re-test this case.

Comment (valerii-bokhan) [reply src/libslic3r/PrintObject.cpp:1404]: I tested separated infills, and it works fine for me. I also added an additional test in the last comment.

**Comment by RF47** [comment]
### Test checklist

* [x] Infill rotation template
* [x] Multiline infill
* [x] Separated infills
* [x] Variable layer height
* [x] Infill combination

**Comment by valerii-bokhan** [comment]
## Infill rotation template = +0.5 , Relative Bridge Angle = enabled

### Infill rotation template = 0

image

### Infill rotation template = 45

Minor bugs can be seen.

image

image

**Comment by valerii-bokhan** [comment]
##  Multiline = 4

### Sparse infill density = 35%, Infill rotation template = 45, Relative Bridge Angle = enabled
image

##  Multiline = 2

### Sparse infill density = 35%, Infill rotation template = 45, Relative Bridge Angle = enabled

image

**Comment by valerii-bokhan** [comment]
## Variable layer height

image

image

image

**Comment by valerii-bokhan** [comment]
## Infill combination = enabled

image

**Comment by valerii-bokhan** [comment]
## Separated infills = enabled

image

See different bridge directions. I'm not sure whether it's a bug.

image

## Separated infills = disabled

image

Bugs:

image

image

**Comment by RF47** [comment]
image
image
image
image
image

**Comment by valerii-bokhan** [comment]
## Thick Internal Bridges = disabled (it's enabled by deafult)

image

**Comment by valerii-bokhan** [comment]
@RF47 I provided screenshots for each item in your checklist - it's mostly working ok, but some bugs are still there. I would look into them in a separate PR, if it's ok.

**Comment by RF47** [comment]
> Thick Internal Bridges = disabled (it's enabled by deafult)

this is an old bug not related to this PR

**Comment by RF47** [comment]
image

**Review by RF47 — APPROVED** [review]
Thanks

**Comment by pi-squared-studio** [comment]
@valerii-bokhan @ianalexis 

It seems that Separated Infills were canceled with this PR.

image

If it helps, in this [PR](https://github.com/OrcaSlicer/OrcaSlicer/pull/15168/changes#diff-824b66f186282a90c303d56c4743be12517ee1b3134e4bf922fef4a353fc280d), in the file src/libslic3r/Fill/FillPlanePath.cpp, I made a rough draft of the debug #define DEBUG_PATH
 — you can use it if needed.

This will help assess the minimum area of the fill, as well as the movement vectors of the polygons.
image

**Comment by valerii-bokhan** [comment]
@pi-squared-studio, can you please attach the test project? Everything worked fine during the tests, so I may have missed some edge cases.

I think it's ' Align directions to model ' that I forgot to test... I will look into it.

**Comment by pi-squared-studio** [comment]
@valerii-bokhan Oh, okay, my bad, everything is working. I just noticed that part of the code related to this was missing in this PR. I didn’t take into account that it might have been optimized only for Assemblies. I think this used to work with models, and I discovered this incongruity when I optimized my PR with the updated branch.

image

**Comment by pi-squared-studio** [comment]
> I think it's ' Align directions to model ' that I forgot to test... I will look into it.

I tested this with the latest branch — everything works.

UPD: when separate infills and align direction, This doesn’t work. But it’s correct in this logic, since the assembly doesn’t contain individual models, but parts.

**Comment by valerii-bokhan** [comment]
## Assembly
image
image

## Separated
image
image

**Comment by valerii-bokhan** [comment]
> > I think it's ' Align directions to model ' that I forgot to test... I will look into it.
> 
> I tested this with the latest branch — everything works.
> 
> UPD: when separate infills and align direction, This doesn’t work. But it’s correct in this logic, since the assembly doesn’t contain individual models, but parts.

I'm glad it's working :)

**Comment by pi-squared-studio** [comment]
@valerii-bokhan Thank you. Sorry for my panic. )))

File: tests/fff_print/test_fill.cpp [hunk 4]
```diff
@@ -1229,3 +1230,68 @@ TEST_CASE("Smoothing multiline lightning infill keeps its outlines connected", "
     REQUIRE(smooth.point_count > sharp.point_count);
     REQUIRE(smooth.sharp_turns < sharp.sharp_turns);
 }
+
+TEST_CASE("Sparse plane-path anchors match the printed infill", "[Fill][InternalBridge][Regression]")
```
Comment (pi-squared-studio) [inline tests/fff_print/test_fill.cpp]: Perhaps this is an unnecessary test, as it does not check the quality of the printed fill.
As far as I understand, he is analyzing whether each line intersects with any other underlying line or the perimeter.

Image

I reproduced the test, and I’d like to note that such a line will fall into the model during printing, as it has no support. But the test will still show a positive result.

I’m having some issues with this test right now because I changed the line rendering for rotating fills https://github.com/OrcaSlicer/OrcaSlicer/pull/15168 (the Hilbert curve used to not rotate around the model’s center). And this test produces an error because there is a free end of the line that does not extend beyond the perimeter. This is similar to the central link in Archimedes’ chord. Also, the aforementioned PR in the new implementation will allow shifting the center, so similar issues that the test won’t detect will still be present, but they will need to be monitored by users.
